8294689: The SA transported_core.html file needs quite a bit of work

plummercj opened the pull request:
The main issues I set out to address are:

- legacy Solaris references can be removed
- add help with dealing with Windows core files, most of important of which is use of PATH to find the dlls referenced by the core file.
- add help with dealing with macOS core files, most of important of which is using JAVA_HOME or DYLD_LIBRARY_PATH to find the dlls referenced by the core file. 

However, in the end it turned out to be a pretty substantial restructuring and rewrite.

Here are links to the rendered old and new copies of the file:
[old file](https://htmlpreview.github.io/?https://github.com/openjdk/jdk/blob/master/src/jdk.hotspot.agent/doc/transported_core.html)
[new file rev1](https://htmlpreview.github.io/?https://github.com/openjdk/jdk/blob/6778a202a855a534754c4b47a6cd99cb912869d3/src/jdk.hotspot.agent/doc/transported_core.html)
[new file rev2](https://htmlpreview.github.io/?https://github.com/openjdk/jdk/blob/7448593f0a9142e5d99844f349b4d54819735540/src/jdk.hotspot.agent/doc/transported_core.html)
[new file rev3](https://htmlpreview.github.io/?https://github.com/openjdk/jdk/blob/ce5c808ca5af4e551bf78d593d865a1f228ad950/src/jdk.hotspot.agent/doc/transported_core.html)
[new file rev4](https://htmlpreview.github.io/?https://github.com/openjdk/jdk/blob/3b16e560499eb526d283a9c16531cc1c3f63394e/src/jdk.hotspot.agent/doc/transported_core.html)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294689](https://bugs.openjdk.org/browse/JDK-8294689): The SA transported_core.html file needs quite a bit of work


### Reviewers
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to [6778a202](https://git.openjdk.org/jdk/pull/10518/files/6778a202a855a534754c4b47a6cd99cb912869d3)
 * [Poonam Bajaj](https://openjdk.org/census#poonam) (@poonamparhar - Committer) ⚠️ Review applies to [3b16e560](https://git.openjdk.org/jdk/pull/10518/files/3b16e560499eb526d283a9c16531cc1c3f63394e)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10518/head:pull/10518` \
`$ git checkout pull/10518`

Update a local copy of the PR: \
`$ git checkout pull/10518` \
`$ git pull https://git.openjdk.org/jdk pull/10518/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10518`

View PR using the GUI difftool: \
`$ git pr show -t 10518`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10518.diff">https://git.openjdk.org/jdk/pull/10518.diff</a>

</details>

kevinjwalls reviewed (approved):
Looks good.  I just looked at SA_ALTROOT and see you can use it not only as a prefix, i.e. the name of a directory which contains a directory structure the same as the original machine.  But you can also use it as a directory which contains the files themselves, without directories.
So both ways you could possibly interpret the text will work.

plummercj commented:
> Looks good. I just looked at SA_ALTROOT and see you can use it not only as a prefix, i.e. the name of a directory which contains a directory structure the same as the original machine. But you can also use it as a directory which contains the files themselves, without directories. So both ways you could possibly interpret the text will work.

I updated the Linux section to better describe SA_ALTROOT. I also updated the macOS section. I had misunderstood parts of it. I noticed this when I looked at SA_ALTROOT a bit closer. For one, I didn't not know that SA_ALTROOT is used  (sort of) on macOS.

poonamparhar commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,87 @@ <h1>Debugging transported core dumps</h1>
-</ul>
+ If the debugger machine and core dump machine have identical Windows libraries, then you only
+ need to point SA to the location of the JDK bin directory. This is done by making sure
+ the bin directory is included in the <b>PATH</b> environment variable. If the windows
```
The location of jvm.dll (e.g. %Java_Home%\bin\server) should be included in the PATH environment variable. And if I remember correctly, the location of 'windbg' should also be added to PATH - this needs to be tested and confirmed though.

plummercj replied:
Ok. So you are saying both the bin and bin\server directories have to be in PATH. I'll update for that. I can't verify the need for windbg. I don't have an environment setup for testing this, but Coleen had indicated it wasn't necessary. Are you setup to try it out?

poonamparhar commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,87 @@ <h1>Debugging transported core dumps</h1>
 When a core dump is moved to a machine different from the one where it was
 produced ("transported core dump"), debuggers (dbx, gdb, windbg or SA) do not
-always successfully open the dump. This is due to kernel, library (shared
-objects or DLLs) mismatch between core dump machine and debugger machine.
```
Looks like an additional space here.

poonamparhar commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,87 @@ <h1>Debugging transported core dumps</h1>
+On  most platforms, core dumps do not contain text (code) pages.
+Their pages are instead  read from the executable and shared objects (or DLLs).
+Therefore it is important to have a matching java executable and shared object
+files on the debugger machine. The best  way to guarantee this match is to match the
```
An extra space between _best_ and _way_.

poonamparhar commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,87 @@ <h1>Debugging transported core dumps</h1>
+debugger machine to that of core dump machine. This means having the same
+OS version and libraries, and also having the same version of the JDK. It also means having
+the OS libraries and JDK installed in the same locations on both machines. However, often this isn't an option,
+and instead you need to include copies of the core dump machine's libraries
```
Perhaps, it can be rephrased as follows:

... instead you need to bring copies of the libraries and java installation from the machine where the core dump was produced to the debugger machine, and ...

plummercj commented:
@poonamparhar  and @kevinjwalls  Can you give one final review since I've made a few minor edits. Thanks:

[new file rev4](https://htmlpreview.github.io/?https://github.com/openjdk/jdk/blob/3b16e560499eb526d283a9c16531cc1c3f63394e/src/jdk.hotspot.agent/doc/transported_core.html)

kevinjwalls commented:
Yes, looks good.
Wouldn't worry much about the double spaces, there are a few others but it's meant to be html, so they don't show in a browser.

poonamparhar commented:
Looks good! Thanks for updating this documentation.

sspitsyn commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,98 @@ <h1>Debugging transported core dumps</h1>
-these debugger libraries can be found in 
-<a href="http://docs.sun.com/app/docs/doc/817-1984/">
-Solaris Linkers and Libraries Guide - 817-1984</a>
+On  most platforms, core dumps do not contain text (code) pages.
```
Nit: an extra space in "On  most".

sspitsyn commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,98 @@ <h1>Debugging transported core dumps</h1>
 
-<h3>Solaris SA against transported core dumps</h3>
-
+<h3>Using  transported core dumps on Linux</h3>
```
Nit: an extra space after "Using".

sspitsyn commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,98 @@ <h1>Debugging transported core dumps</h1>
+ <b>/altroot/lib/libfoo.so</b>, and <b>/altroot/libfoo.so</b>.
 </p>
+
+<h3>Using  transported core dumps on Windows</h3>
```
Nit: an extra space after "Using".

sspitsyn commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,98 @@ <h1>Debugging transported core dumps</h1>
+ You can also use <b>SA_ALTROOT</b> to specify a path mapping. For example, if you
+ set <b>SA_ALTROOT=/altroot</b>, then <b>/altroot</b> will be prepended to any path found
+ in the core file, and also prepended to any subdir with the root part stripped off. For example,
+ when looking up <b>/usr/lib/libfoo.so</b>, SA will try to find <b>/altroot/usr/lib/libfoo.so</b>,
```
Nit: Not sure if it makes sense to say: "SA will first try to find".

plummercj replied:
I decided to keep this as-is since it will try to find all 3, and I felt saying `first...and then...and finally...` was a bit wordy.

sspitsyn replied:
Okay.

sspitsyn commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,98 @@ <h1>Debugging transported core dumps</h1>
+to the debugger machine and included in <b>PATH</b>.
 </p>
 
+<h3>Using  transported core dumps on macOS</h3>
```
Nit: an extra space after "Using".

sspitsyn commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,98 @@ <h1>Debugging transported core dumps</h1>
-Note that this version of libproc.so has a special symbol to support transported core dump debugging. 
-In future, we may get this feature built into standard libproc.so -- if that happens, this step (of 
-setting LD_LIBRARY_PATH) can be skipped.
+SA normally uses the path to the specified java executable to locate the JDK libraries. It will look in the following subdirectories for them (relative to the path to the specified java executable): <b>../lib</b>, <b>../lib/server</b>, <b>../jre/lib</b>, and <b>../jre/lib/server</b>. If not found in any of those locations, it will look in the same subdirectories relative to the <b>JAVA_HOME</b> e ...
```
Nit: too long line - hard to review.

sspitsyn commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,98 @@ <h1>Debugging transported core dumps</h1>
-set, SA ignores libthread_db failure, but you won't be able to get any
-thread related information. But, you would be able to use SA and get
-other information.
+For locating the user JNI libraries, SA uses <b>DYLD_LIBRARY_PATH</b>. It can contain more than one directory separated by a colon. <b>DYLD_LIBRARY_PATH</b> can also be used for locating the JDK libraries, but it needs to specify the full path to the libraries. SA will not automatically search subdirs such as <b>lib/server</b> as it does for <b>JAVA_HOME</b>.
```
Nit: too long line - hard to review.

sspitsyn commented on src/jdk.hotspot.agent/doc/transported_core.html:
```diff
@@ -10,101 +10,98 @@ <h1>Debugging transported core dumps</h1>
-workaround this, you can define environment variable <b>SA_ALTROOT</b>
-to be the directory where shared libraries are kept. The semantics of
-this env. variable is same as that for Solaris (please refer above).
+For locating the macOS libraries, SA uses <b>SA_ALTROOT</b> similar to the linux support, except it does not use it to map all the subdirs. It just appends <b>SA_ALTROOT</b> to the full path of each macOS library. So if you specify <b>SA_ALTROOT=/altroot</b>, SA will prepend <b>/altroot</b> to the full path of each macOS library. Note however, due to <a href="https://bugs.openjdk.org/browse/JDK-8 ...
```
Nit: too long line - hard to review.

plummercj replied:
Ok, I split these all up into shorter lines. The rendered html is still the same.

sspitsyn reviewed (approved):
Looks good to me.
I've posted several nits though.
Thanks,
Serguei
